crates/storage/provider/src/post_state.rs

```diff
@@ -0,0 +1,828 @@
+        mut self,
+        tx: &TX,
+        first_transition_id: TransitionId,
+    ) -> Result<(), DbError> {
+        // Collect and sort changesets by their key to improve write performance
+        let mut changesets = std::mem::take(&mut self.changes);
+        changesets
+            .sort_unstable_by_key(|changeset| (changeset.transition_id(), changeset.address()));
+
+        // Partition changesets into account and storage changes
+        let (account_changes, storage_changes): (Vec<Change>, Vec<Change>) =
+            changesets.into_iter().partition(|changeset| {
+                matches!(
+                    changeset,
+                    Change::AccountChanged { .. } |
+                        Change::AccountCreated { .. } |
+                        Change::AccountDestroyed { .. }
+                )
+            });
+
+        // Write account changes
+        let mut account_changeset_cursor = tx.cursor_dup_write::<tables::AccountChangeSet>()?;
+        for changeset in account_changes.into_iter() {
+            match changeset {
+                Change::AccountDestroyed { id, address, old } |
+                Change::AccountChanged { id, address, old, .. } => {
+                    account_changeset_cursor.append_dup(
+                        first_transition_id + id,
+                        AccountBeforeTx { address, info: Some(old) },
+                    )?;
+                }
+                Change::AccountCreated { id, address, .. } => {
+                    account_changeset_cursor.append_dup(
+                        first_transition_id + id,
+                        AccountBeforeTx { address, info: None },
+                    )?;
+                }
+                _ => unreachable!(),
+            }
+        }
+
+        // Write storage changes
+        let mut storages_cursor = tx.cursor_dup_write::<tables::PlainStorageState>()?;
+        let mut storage_changeset_cursor = tx.cursor_dup_write::<tables::StorageChangeSet>()?;
+        for changeset in storage_changes.into_iter() {
+            match changeset {
+                Change::StorageChanged { id, address, changeset } => {
+                    let storage_id = TransitionIdAddress((first_transition_id + id, address));
+
+                    for (key, (old_value, _)) in changeset {
+                        storage_changeset_cursor.append_dup(
+                            storage_id,
+                            StorageEntry { key: H256(key.to_be_bytes()), value: old_value },
+                        )?;
+                    }
+                }
+                Change::StorageWiped { id, address } => {
+                    let storage_id = TransitionIdAddress((first_transition_id + id, address));
+
+                    if let Some((_, entry)) = storages_cursor.seek_exact(address)? {
+                        storage_changeset_cursor.append_dup(storage_id, entry)?;
+
+                        while let Some(entry) = storages_cursor.next_dup_val()? {
```

Comment:
Excellent - cc @joshieDo on usage of the dup apis which you also made use of in the hashing/merkle stages and seem to be impactful